cmd/repo-updater/internal/authz/integration_test.go

```diff
@@ -489,3 +491,153 @@ func TestIntegration_GitHubPermissions(t *testing.T) {
+		},
+		{
+			Name:    "gitlab.sgdev.org/petri.last/getschwifty",
+			Private: true,
+			URI:     "gitlab.sgdev.org/petri.last/getschwifty",
+			ExternalRepo: api.ExternalRepoSpec{
+				ID:          "371334",
+				ServiceType: extsvc.TypeGitLab,
+				ServiceID:   "https://gitlab.sgdev.org/",
+			},
+			Sources: map[string]*types.SourceInfo{
+				svc.URN(): {
+					ID: svc.URN(),
+				},
+			},
+		},
+	}
+
+	authData := json.RawMessage(fmt.Sprintf(`{"access_token": "%s"}`, token))
+
+	// This integration tests performs a user-centric permissions syncing against
+	// https://gitlab.sgdev.org, then check if permissions are correctly granted for the test
+	// user "sourcegraph-vcr".
+	t.Run("test feature flag", func(t *testing.T) {
+		name := t.Name()
+
+		cf, save := httptestutil.NewRecorderFactory(t, update(name), name)
+		defer save()
+		doer, err := cf.Doer()
+		require.NoError(t, err)
+
+		testDB := database.NewDB(logger, dbtest.NewDB(logger, t))
+
+		ctx := actor.WithInternalActor(context.Background())
+
+		reposStore := repos.NewStore(logtest.Scoped(t), testDB)
+
+		err = reposStore.ExternalServiceStore().Upsert(ctx, &svc)
+		require.NoError(t, err)
+
+		provider := authzGitLab.NewOAuthProvider(authzGitLab.OAuthProviderOp{
+			BaseURL: uri,
+			DB:      testDB,
+			CLI:     doer,
+		})
+
+		authz.SetProviders(false, []authz.Provider{provider})
+		defer authz.SetProviders(true, nil)
+		for _, repo := range testRepos {
+			err = reposStore.RepoStore().Create(ctx, &repo)
+			require.NoError(t, err)
+		}
+
+		user, err := testDB.UserExternalAccounts().CreateUserAndSave(ctx, newUser, spec, extsvc.AccountData{
+			AuthData: extsvc.NewUnencryptedData(authData),
+		})
+		require.NoError(t, err)
+
+		permsStore := database.Perms(logger, testDB, timeutil.Now)
+		syncer := NewPermsSyncer(logger, testDB, reposStore, permsStore, timeutil.Now)
+
+		assertUserPermissions := func(t *testing.T, wantIDs []int32) {
+			t.Helper()
+			_, providerStates, err := syncer.syncUserPerms(ctx, user.ID, false, authz.FetchPermsOptions{})
+			require.NoError(t, err)
+
+			assert.Equal(t, database.CodeHostStatusesSet{{
+				ProviderID:   "https://gitlab.sgdev.org/",
+				ProviderType: "gitlab",
+				Status:       database.CodeHostStatusSuccess,
+				Message:      "FetchUserPerms",
+			}}, providerStates)
+
+			p, err := permsStore.LoadUserPermissions(ctx, user.ID)
+			require.NoError(t, err)
+
+			gotIDs := make([]int32, len(p))
+			for i, perm := range p {
+				gotIDs[i] = perm.RepoID
+			}
+
+			if diff := cmp.Diff(wantIDs, gotIDs); diff != "" {
+				t.Fatalf("IDs mismatch (-want +got):\n%s", diff)
+			}
+		}
+
+		t.Run("feature-flag disabled", func(t *testing.T) {
+			assertUserPermissions(t, []int32{2})
+		})
+
+		t.Run("feature-flag enabled", func(t *testing.T) {
+			_, err = testDB.FeatureFlags().CreateBool(ctx, "gitLabProjectVisibilityExperimental", true)
+			require.NoError(t, err, "feature flag creation failed")
+
+			assertUserPermissions(t, []int32{1, 2})
+		})
```

Comment:
I'd personally remove the `t.Run` here. `t.Run` always suggests to me that these tests are independent and can be run independently. But in this case, if the 2nd test runs before the first one, the first one will fail, because "global" state has been modified.

So I'd personally do something like this:


```suggestion
		// Without the featuer flag, we should only get one repository back:
		assertUserPermissions(t, []int32{2})

		// But with the feature flag enabled, user should have access to both of them:
		_, err = testDB.FeatureFlags().CreateBool(ctx, "gitLabProjectVisibilityExperimental", true)
		require.NoError(t, err, "feature flag creation failed")

		assertUserPermissions(t, []int32{1, 2})
```

That makes it clearer, IMHO, that the tests are dependent on the same state and that the order is important.